changefeedccl: Introduce exactly once semantics for changefeed export

miretskiy opened the pull request:
Changefeed supports export mode where the table (or tables) may be exported to any of the supported changefeed sinks by specifying `initial_scan=only` option.

Changefeed exports are preferable to the `EXPORT` command particularly when exporting large amouints of data. The reason for this is that changefeed checkpoints its progress, whereas `EXPORT` command restarts from the begining in case of a transient error.

Due to the possibility of restarts from checkpoint, the changefeed exports may produce duplicate entries for the same key.

This PR introduces a new option `exactly_once` that can be specified if the duplicates are undesirable. This option is supported for file based sinks only (s3, gcp, azure, etc).

Informs #105297

Technical description of the change:
The changefeed with exactly_once option begins execution like any
other changefeed. Prior to commiting job record for the changefeed,
an execution plan is produced, specifying the distribution of the
ranges across nodes. This execution plan is persisted into job record.
The static partitionining specifies the list of spans that belong
to a partition, and it is used instead of the regular partitioning
scheme when changefeed starts to run.

When static partitioning is set, those partition are assigned to
all nodes that match optional locality in a round-robbin fashion.
That means, for example, that if the "shape" of the cluster is changed
(nodes removed/added), partitions may be assigned to different
nodes or even multiple partition will be assigned to multiple
aggregators on a single node.

It is important to understand how checkpoints are managed when
running `initial_scan=only` export.  Initially, the "frontier" timestamp
is set to 0.  The changefeed will scan all data as of "statement
timestamp" -- that is, we will read all data as of the changefeed
creation time.  Each aggregator will be assigned a portion of the
overall workload.  Aggregator will begin scanning those ranges, and,
as the data arrives, the data is processed and written out to an
output sink.  File based output sinks may buffer the data, and will
flush the data out (i.e. write the file) if the buffer becomes too
large.  These are called size based flushes.

Critically, whenever aggregator completes exporting
a single range, it will emit a record indicating that  that range
is done -- and this ranges timestamp is now equal to statement
timestamp.  Periodically (`changefeed.frontier_checkpoint_frequency` --
10 minutes by default), change aggregator will want to forward its
current checkpoint information (essentially, the list of ranges
and their timestamps) to the coordinator.  *Prior* to
forwarding checkpoint information to the change coordinator, the
aggregator will force flush all buffered data to ensure that any
in-flight data is written out before checkpoint information is
sent to the coordinator.  These are called explicit flushes.
The flush generation is incremented, and is sent to the coordinator.

The coordinator unifies aggregator checkpoints into a global checkpoint,
and When all ranges reach statemet timestamp, the frontier
advances to this position, and the changefeed terminates.

The cloud storage sink is modified so that the each file produced
by the sink is prefixed with the `<partition id>~<flush generation>`.
Both parts are encoded as a 0 prefixed hexadecimal string (`%08x`).

Focusing on a single aggregator which is assigned a single partition
identified by its ordinal number in the static partitioning (i.e. it's
just an integer).  The aggregator will arrange to start the scan
of the ranges assigned to it.  Every time an explicit flush is issued
(i.e. right before aggregator wishes to send its checkpoint to the
coordinator), a "flush generation" counter is incremented.
This counter is communicated to the coordinator, along with the
partition ID.  The coordinator unifies all of these flush counters
to keep track of overal progress over entire static partition.

When the coordinator writes out checkpoint, it ensures that the
entire frontier checkpoint is written out, and that flush generation
sequences are persisted for each partition.  This guarantees that
for all aggregators, we know that all files with flush generation
less than checkpointed flush generation, are correct and do not contain
duplicates (we know that there are no duplicates because data Scan
operation used to process initial scan does not produce duplicates, as
opposed to rangefeed).

Upon restart due to a transient error, changefeed aggregators will
remove all files generated by the partition assigned to this aggregator,
that have flush generation that's equal or is higher to the last
checkpoint flush generation.  Some of these removals could be avoided,
but it's just more expensive to try to re-read those files to reproduce
exact and correct list of ranges to scan.

Release note (enterprise change): Changefeed exports (that is changefeeds that run with `initial_scan=only` option) now support new `exactly_once` option for file based sinks which, when specified, ensures that the exported data will not contain any duplicates for the same key.

Commit:
changefeedccl: Introduce exactly once semantics for changefeed export

Changefeed supports export mode where the table (or tables) may be
exported to any of the supported changefeed sinks by specifying
`initial_scan=only` option.

Changefeed exports are preferable to the `EXPORT` command
particularly when exporting large amouints of data.
The reason for this is that changefeed checkpoints its progress,
whereas `EXPORT` command restarts from the begining in case
of a transient error.

Due to the possibility of restarts from checkpoint, the changefeed
exports may produce duplicate entries for the same key.

This PR introduces a new option `exactly_once` that
can be specified when if the duplicates are undesirable.
This option is supported for file based sinks only (s3, gcp, azure, etc).

Informs #105297

Technical description of the change:
The changefeed with exactly_once option begins execution like any
other changefeed. Prior to commiting job record for the changefeed,
an execution plan is produced, specifying the distribution of the
ranges across nodes. This execution plan is persisted into job record.
The static partitionining specifies the list of spans that belong
to a partition, and it is used instead of the regular partitioning
scheme when changefeed starts to run.

When static partitioning is set, those partition are assigned to
all nodes that match optional locality in a round-robbin fashion.
That means, for example, that if the "shape" of the cluster is changed
(nodes removed/added), partitions may be assigned to different
nodes or even multiple partition will be assigned to multiple
aggregators on a single node.

It is important to understand how checkpoints are managed when
running `initial_scan=only` export.  Initially, the "frontier" timestamp
is set to 0.  The changefeed will scan all data as of "statement
timestamp" -- that is, we will read all data as of the changefeed
creation time.  Each aggregator will be assigned a portion of the
overall workload.  Aggregator will begin scanning those ranges, and,
as the data arrives, the data is processed and written out to an
output sink.  File based output sinks may buffer the data, and will
flush the data out (i.e. write the file) if the buffer becomes too
large.  These are called size based flushes.

Critically, whenever aggregator completes exporting
a single range, it will emit a record indicating that  that range
is done -- and this ranges timestamp is now equal to statement
timestamp.  Periodically (`changefeed.frontier_checkpoint_frequency` --
10 minutes by default), change aggregator will want to forward its
current checkpoint information (essentially, the list of ranges
and their timestamps) to the coordinator.  *Prior* to
forwarding checkpoint information to the change coordinator, the
aggregator will force flush all buffered data to ensure that any
in-flight data is written out before checkpoint information is
sent to the coordinator.  These are called explicit flushes.
The flush generation is incremented, and is sent to the coordinator.

The coordinator unifies aggregator checkpoints into a global checkpoint,
and When all ranges reach statemet timestamp, the frontier
advances to this position, and the changefeed terminates.

The cloud storage sink is modified so that the each file produced
by the sink is prefixed with the `<partition id>~<flush generation>`.
Both parts are encoded as a 0 prefixed hexadecimal string (`%08x`).

Focusing on a single aggregator which is assigned a single partition
identified by its ordinal number in the static partitioning (i.e. it's
just an integer).  The aggregator will arrange to start the scan
of the ranges assigned to it.  Every time an explicit flush is issued
(i.e. right before aggregator wishes to send its checkpoint to the
coordinator), a "flush generation" counter is incremented.
This counter is communicated to the coordinator, along with the
partition ID.  The coordinator unifies all of these flush counters
to keep track of overal progress over entire static partition.

When the coordinator writes out checkpoint, it ensures that the
entire frontier checkpoint is written out, and that flush generation
sequences are persisted for each partition.  This guarantees that
for all aggregators, we know that all files with flush generation
less than checkpointed flush generation, are correct and do not contain
duplicates (we know that there are no duplicates because data Scan
operation used to process initial scan does not produce duplicates, as
opposed to rangefeed).

Upon restart due to a transient error, changefeed aggregators will
remove all files generated by the partition assigned to this aggregator,
that have flush generation that's equal or is higher to the last
checkpoint flush generation.  Some of these removals could be avoided,
but it's just more expensive to try to re-read those files to reproduce
exact and correct list of ranges to scan.

Release note (enterprise change): Changefeed exports (that is
changefeeds that run with `initial_scan=only` option) now support
new `exactly_once` option for file based sinks which, when specified,
ensures that the exported data will not contain any duplicates for the
same key.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/105764)
<!-- Reviewable:end -->

jayshrivastava reviewed (changes requested):
I did one pass and left some comments. This is a nice PR and the strategy makes sense. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-:-N_jsEuj1Qxb8YB0cQjB:b1ncmxk)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @rhu713, and @samiskin)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 392 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_ji9BtE2SB9HNPBV28:-N_ji9BtE2SB9HNPBV29:b5l9x15) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/changefeed_dist.go#L392)):*
> ```Go
> 
> 	var spanPartitions []sql.SpanPartition
> 	if len(details.StaticPartition) == 0 {
> ```

nit:  having a method "hasStaticPartition" on the details might be cleaner
___
*[`pkg/ccl/changefeedccl/sink_cloudstorage.go` line 659 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_jo6Lx8g1aoHZfVNwz:-N_jo6Lx8g1aoHZfVNx-:brj56m2) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/sink_cloudstorage.go#L659)):*
> ```Go
> 
> 	if s.exactlyOnce.enabled {
> 		s.exactlyOnce.flushGen++
> ```

Maintaining two copies of and incrementing them in lockstep is a little dangerous. I feel like there is some potential for bugs there - if not now, then in the future. 
___
*[`pkg/ccl/changefeedccl/sink_cloudstorage.go` line 879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_jmF2D7Joe9HBiND-Q:-N_jmF2D7Joe9HBiND-R:b9ohlvc) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/sink_cloudstorage.go#L879)):*
> ```Go
> 	s.exactlyOnce.flushGen = flushGen
> 
> 	if flushGen == 0 {
> ```

A comment would be nice here. 
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 1091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_joiHfDzMcrXUIea9O:-N_joiHfDzMcrXUIea9P:b-pzwe25) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/changefeedbase/options.go#L1091)):*
> ```Go
> 	}
> 
> 	if s.IsSet(OptExactlyOnce) {
> ```

Maybe check for the cloudstorage sink here too? 
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7341 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_jenZx4Vj5ikKflf4v:-N_jenZx4Vj5ikKflf4w:b-ebgsv0) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/changefeed_test.go#L7341)):*
> ```Go
> }
> 
> func TestChangefeedExactlyOnceExport(t *testing.T) {
> ```

Add test using CDC queries? Just to check the partitioning/planning logic? If it's not worth it, then don't worry.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_jrUMlB4Ti5LTMkO88:-N_jrUMm1ouZJMU3ANJx:b-ue73et) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/changefeed_test.go#L7342)):*
> ```Go
> 
> func TestChangefeedExactlyOnceExport(t *testing.T) {
> 	defer leaktest.AfterTest(t)()
> ```

There's a few more things we can assert for more test coverage:
- The changefeed retries at a point such that we know for sure at least one previously written file must be removed by the cloudstorage sink to ensure no duplicates
- Filenames are correct (the encoded flush generation number is different before and after the retry)


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105764)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava, @rhu713, and @samiskin)

___
*[`pkg/ccl/changefeedccl/sink_cloudstorage.go` line 659 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_jo6Lx8g1aoHZfVNwz:-NawQFkrDHK-np3YPXuW:bkmfpse) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/sink_cloudstorage.go#L659)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Maintaining two copies of and incrementing them in lockstep is a little dangerous. I feel like there is some potential for bugs there - if not now, then in the future. 
</blockquote></details>

yes; i didn't like it myself.  Fixed by adding a small interface which is instantiated in the aggregator, and used here.  just like "timestamp oracle".
___
*[`pkg/ccl/changefeedccl/sink_cloudstorage.go` line 879 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_jmF2D7Joe9HBiND-Q:-NawQ_5a20mdwE7LYHZ3:bukygi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/sink_cloudstorage.go#L879)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

A comment would be nice here. 
</blockquote></details>

Nice one.  Cause of course, once I started writing the comment, I realized that this check is just wrong.
The intention was: don't do anything else if this is the first time we are starting export.
But that's just wrong: we could start with flushGen==0; then write bunch of files, and then crash.
Those "0 generation" files should be deleted.

Added a test for this.
Thanks.
___
*[`pkg/ccl/changefeedccl/changefeedbase/options.go` line 1091 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_joiHfDzMcrXUIea9O:-NawVC3A2AxdNieFz3Rt:bqktyn3) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/changefeedbase/options.go#L1091)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Maybe check for the cloudstorage sink here too? 
</blockquote></details>

Can't do it here -- but I added checks when we create new sink, as well as inside tryEnableExactlyOnceSemantics.
Let me know what you think.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7341 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_jenZx4Vj5ikKflf4v:-Nawcene2i1FKEPsxEPA:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/changefeed_test.go#L7341)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Add test using CDC queries? Just to check the partitioning/planning logic? If it's not worth it, then don't worry.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_jrUMlB4Ti5LTMkO88:-NawVPaM9INR5CwPHRCC:b-ruticg) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/changefeed_test.go#L7342)):*
> the encoded flush generation number is different before and after the retry
What do you mean? Retries should use old (i.e. checkpointed) generation numbers and the file names
should be the same...

* The changefeed retries at a point such that we know for sure at least one previously written file must be removed by the cloudstorage sink to ensure no duplicates
*
The way the test setup, it definitely has to remove the files -- because I force file flush on every resolved span, and that after seen few of those resolved spans, i fail the changefeed -- all previous files must have been removed; otherwise, the test fails because it detects duplicates.


<!-- Sent from Reviewable.io -->

miretskiy commented:
@jayshrivastava -- thanks for the review;
I resurrected/rebased/fixed this change, and addressed all your comments.
Not sure if we'll merge this PR, but please review it.

jayshrivastava commented:
@miretskiy I'll take a look tomorrow!

jayshrivastava reviewed (approved):
LGTM! I left some small comments. Feel free to disregard any you disagree with.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-:-Nb06l_S1nkQ7f0gL9jY:b-hd905k)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @miretskiy, @rhu713, and @samiskin)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 325 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nb04hy23Q53EfZ1O3AW:-Nb04hy23Q53EfZ1O3AX:b-crgnep) ([raw file](https://github.com/cockroachdb/cockroach/blob/33bd35040d8d60087361bf6496030227782c4772/pkg/ccl/changefeedccl/changefeed_dist.go#L325)):*
> ```Go
> }
> 
> // defaultSpanPartition partitions spans using default approach.
> ```

nit: comment name doesnt match the func name
Also maybe to differentiate it between jobspb.ChangefeedDetails_Partition, which does not pin partitions to nodes, you can call this function `staticNodePartitions`.
___
*[`pkg/ccl/changefeedccl/changefeed_processors.go` line 793 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nb01Yc_8EcKbPLhz2lP:-Nb01Yca9oq6puOZOJxa:b4zxn3i) ([raw file](https://github.com/cockroachdb/cockroach/blob/33bd35040d8d60087361bf6496030227782c4772/pkg/ccl/changefeedccl/changefeed_processors.go#L793)):*
> ```Go
> 	progressUpdate := jobspb.ChangeAggregatorProgress{
> 		ResolvedSpans:   batch,
> 		Partition:       ca.spec.Partition,
> ```

It may be worth noting in a comment that these are zero-values if exactly_once is not specified and that they are ignored by the coordinator if exactly_once is not specified. This also makes it clear to the reader:
```
if ca.ExactlyOnce() {
    progressUpdate.Partition =   ca.spec.Partition
     progressUpdate.FlushGeneration =  ca.flushGeneration.get()
}
```
How do you feel about starting the FlushGeneration number at 1 instead of 0 in the case what exactly_once is actually enabled? 
___
*[`pkg/ccl/changefeedccl/changefeed_processors.go` line 1048 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nb00SWwEvWDAlE9ICQL:-Nb00SWwEvWDAlE9ICQM:b-3rcbmu) ([raw file](https://github.com/cockroachdb/cockroach/blob/33bd35040d8d60087361bf6496030227782c4772/pkg/ccl/changefeedccl/changefeed_processors.go#L1048)):*
> ```Go
> 	opts := changefeedbase.MakeStatementOptions(cf.spec.Feed.Opts)
> 	cf.exactlyOnce.enabled = opts.ExactlyOnceExport()
> 	cf.exactlyOnce.flushProgress = make([]int64, len(spec.Feed.StaticPartition))
> ```

we don't need to make this unless it is enabled
___
*[`pkg/ccl/changefeedccl/sink.go` line 201 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nb-zTva-2qV0LqBCKg0:-Nb-zTva-2qV0LqBCKg1:b-53ivrh) ([raw file](https://github.com/cockroachdb/cockroach/blob/33bd35040d8d60087361bf6496030227782c4772/pkg/ccl/changefeedccl/sink.go#L201)):*
> ```Go
> 	opts := changefeedbase.MakeStatementOptions(feedCfg.Opts)
> 
> 	if opts.ExactlyOnceExport() && !sinkSupportsExactlyOnceExport(u.Scheme) {
> ```

This might be dead code and we may not need `sinkSupportsExactlyOnceExport`. We call `validateSinkOptions` below and you've added this option to `CloudStorageValidOptions`, so one would think that the validation code does this check for you. Even if it isn't dead code, putting this option in `CloudStorageValidOptions` _should_ make validation fail without this code. 
___
*[`pkg/jobs/jobspb/jobs.proto` line 1096 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nb-y2PqAqMMrd2K-qzF:-Nb-y2PqAqMMrd2K-qzG:b-cyaxcm) ([raw file](https://github.com/cockroachdb/cockroach/blob/33bd35040d8d60087361bf6496030227782c4772/pkg/jobs/jobspb/jobs.proto#L1096)):*
> ```proto
>   ];
> 
>   // Flush generation for a static partition used during changefeed EXPORT.
> ```

nit: it would be nice to include that the length of this matches the number of static partitions in the changefeed details struct
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7342 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-N_jrUMlB4Ti5LTMkO88:-Nb06dPnDHhRSe5VMwjm:bb6sh25) ([raw file](https://github.com/cockroachdb/cockroach/blob/5f6b7e89fc07f48539852969ca40361216e541bf/pkg/ccl/changefeedccl/changefeed_test.go#L7342)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

> the encoded flush generation number is different before and after the retry
What do you mean? Retries should use old (i.e. checkpointed) generation numbers and the file names
should be the same...

* The changefeed retries at a point such that we know for sure at least one previously written file must be removed by the cloudstorage sink to ensure no duplicates
*
The way the test setup, it definitely has to remove the files -- because I force file flush on every resolved span, and that after seen few of those resolved spans, i fail the changefeed -- all previous files must have been removed; otherwise, the test fails because it detects duplicates.
</blockquote></details>

Ack!


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105764)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava, @rhu713, and @samiskin)

___
*[`pkg/ccl/changefeedccl/changefeed_dist.go` line 325 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nb04hy23Q53EfZ1O3AW:-Nb094_MAFqH3bUm7CoT:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33bd35040d8d60087361bf6496030227782c4772/pkg/ccl/changefeedccl/changefeed_dist.go#L325)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

nit: comment name doesnt match the func name
Also maybe to differentiate it between jobspb.ChangefeedDetails_Partition, which does not pin partitions to nodes, you can call this function `staticNodePartitions`.
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/changefeed_processors.go` line 1048 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nb00SWwEvWDAlE9ICQL:-Nb09Xvd5kLSsuZtTHXm:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/33bd35040d8d60087361bf6496030227782c4772/pkg/ccl/changefeedccl/changefeed_processors.go#L1048)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

we don't need to make this unless it is enabled
</blockquote></details>

Done.
___
*[`pkg/ccl/changefeedccl/sink.go` line 201 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nb-zTva-2qV0LqBCKg0:-Nb09evM3ksV1iD7UIDJ:b-t8zqc6) ([raw file](https://github.com/cockroachdb/cockroach/blob/33bd35040d8d60087361bf6496030227782c4772/pkg/ccl/changefeedccl/sink.go#L201)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

This might be dead code and we may not need `sinkSupportsExactlyOnceExport`. We call `validateSinkOptions` below and you've added this option to `CloudStorageValidOptions`, so one would think that the validation code does this check for you. Even if it isn't dead code, putting this option in `CloudStorageValidOptions` _should_ make validation fail without this code. 
</blockquote></details>

No, I don't think so.  I don't want to support this for all cloud storage options.  In particular, I'm excluding http sink -- whilch, while technically could work, in practice it almost certainly won't.
So, this is a bit tighter check; and you know, having extra check doesn't really hurt.  So, i'd like to keep it.


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (approved):
I gave it another once over. I think it would be nice to start baking this in master. Maybe we can add a roachtest in the future once we have more correctness testing infrastructure?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-:-NlA3-eS8VO-e6qjLpSH:bbhmrce)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @miretskiy)

___
*[`-- commits` line 32 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nl9zePA6i1kOWkXUI1I):*
It might make sense to throw an error or something if the number of aggregators is too small compared to the number of partitions. You won't want things to be too imbalanced :) We can just file an issue for this case and decide what to do later.
___
*[`pkg/ccl/changefeedccl/changefeed_processors.go` line 1461 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-NlA1LmM8hqu_GLzNW0t:-NlA1LmM8hqu_GLzNW0u:b-dxv1kf) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ca19c685fb588071e607eae908d6171dfe4e6f0/pkg/ccl/changefeedccl/changefeed_processors.go#L1461)):*
> ```Go
> 		}
> 
> 		forceSpanCheckpoint := cf.knobs.AlwaysCheckpointSpans && i == len(progress.ResolvedSpans.ResolvedSpans)-1
> ```

Could use a comment about why `i == len(progress.ResolvedSpans.ResolvedSpans)-1` is included

Also, some of the logic is scattered here. `forwardFrontier` calls `maybeCheckpointJob` and these functions (and whatever functions they call) do some logic to decide if they want to checkpoint. It's all a bit. Would be nice if it was one if statement ideally. We can address in the future though. For now, this is okay. 


<!-- Sent from Reviewable.io -->

miretskiy reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105764)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @jayshrivastava and @michae2)

___
*[`-- commits` line 32 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nl9zePA6i1kOWkXUI1I):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

It might make sense to throw an error or something if the number of aggregators is too small compared to the number of partitions. You won't want things to be too imbalanced :) We can just file an issue for this case and decide what to do later.
</blockquote></details>

I think it's fine...the reality is that you would need to rerun your export
if things are too slow (or if the cluster size changed drastically).
___
*[`pkg/ccl/changefeedccl/changefeed_processors.go` line 793 at r3](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nb01Yc_8EcKbPLhz2lP:-Nl_MEgl6S3tZmasTPvi:b2e9ocl) ([raw file](https://github.com/cockroachdb/cockroach/blob/33bd35040d8d60087361bf6496030227782c4772/pkg/ccl/changefeedccl/changefeed_processors.go#L793)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

It may be worth noting in a comment that these are zero-values if exactly_once is not specified and that they are ignored by the coordinator if exactly_once is not specified. This also makes it clear to the reader:
```
if ca.ExactlyOnce() {
    progressUpdate.Partition =   ca.spec.Partition
     progressUpdate.FlushGeneration =  ca.flushGeneration.get()
}
```
How do you feel about starting the FlushGeneration number at 1 instead of 0 in the case what exactly_once is actually enabled? 
</blockquote></details>

Good idea; 

I don't feel that we need to start flush gen at 1.  Going to keep it as is,  unless you feel strongly.
___
*[`pkg/ccl/changefeedccl/changefeed_processors.go` line 1461 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-NlA1LmM8hqu_GLzNW0t:-Nl_Mfd83T5RCNZNdBYM:b-qcxdpo) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ca19c685fb588071e607eae908d6171dfe4e6f0/pkg/ccl/changefeedccl/changefeed_processors.go#L1461)):*
<details><summary><i>Previously, jayshrivastava (Jayant) wrote…</i></summary><blockquote>

Could use a comment about why `i == len(progress.ResolvedSpans.ResolvedSpans)-1` is included

Also, some of the logic is scattered here. `forwardFrontier` calls `maybeCheckpointJob` and these functions (and whatever functions they call) do some logic to decide if they want to checkpoint. It's all a bit. Would be nice if it was one if statement ideally. We can address in the future though. For now, this is okay.
</blockquote></details>

comment added.


<!-- Sent from Reviewable.io -->

jayshrivastava reviewed (changes requested):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/105764)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @michae2 and @miretskiy)

___
*[`pkg/ccl/changefeedccl/changefeed_processors.go` line 1461 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-NlA1LmM8hqu_GLzNW0t:-Nl_S3cW8ui1Lopuo1Vh:bhbzqbc) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ca19c685fb588071e607eae908d6171dfe4e6f0/pkg/ccl/changefeedccl/changefeed_processors.go#L1461)):*
<details><summary><i>Previously, miretskiy (Yevgeniy Miretskiy) wrote…</i></summary><blockquote>

comment added.
</blockquote></details>

I see. It was for testing.
___
*[`pkg/ccl/changefeedccl/changefeed_processors.go` line 1569 at r4](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nl_Sf5iDZq7KrHevj2k:-Nl_Sf5iDZq7KrHevj2l:b-jprk89) ([raw file](https://github.com/cockroachdb/cockroach/blob/7ca19c685fb588071e607eae908d6171dfe4e6f0/pkg/ccl/changefeedccl/changefeed_processors.go#L1569)):*
> ```Go
> 		checkpoint.Spans, checkpoint.Timestamp = cf.frontier.getCheckpointSpans(maxBytes)
> 	}
> 	updateCheckpoint = updateCheckpoint || forceSpanCheckpoint
> ```

Is this a mistake? It seems that  `forceSpanCheckpoint` doesn't do anything once you read the testing knob.
___
*[`pkg/ccl/changefeedccl/changefeed_test.go` line 7461 at r5](https://reviewable.io/reviews/cockroachdb/cockroach/105764#-Nl_Tc9v-6-bxE4vKMHK:-Nl_Tc9v-6-bxE4vKMHL:b2w3krl) ([raw file](https://github.com/cockroachdb/cockroach/blob/ea387a04a01ed8436b222683c2088a477223f3fa/pkg/ccl/changefeedccl/changefeed_test.go#L7461)):*
> ```Go
> 	args := base.TestClusterArgs{
> 		ServerArgs: base.TestServerArgs{
> 			DefaultTestTenant: base.TestIsSpecificToStorageLayerAndNeedsASystemTenant, // need nodelocal and splits.
> ```

I used to think that splitting didn't work on non system tenants, but it does. There's still old tests which no one updated with comments that say it doesn't work. Someone fixed it at some point, but didn't update those old tests. Nodelocal might work too. I think it's worth checking for extra test coverage. 

See https://github.com/cockroachdb/cockroach/pull/115166/files `newRangeDistributionTester`. I wrote the whole thing with TODOTenantDisabled, but ended up removing that. It wasn't too much work.


<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
[![CLA assistant check](https://cla.crdb.dev/pull/badge/not_signed)](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=105764) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.crdb.dev/cockroachdb/cockroach?pullRequest=105764) before we can accept your contribution.<br/><hr/>**Yevgeniy Miretskiy** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.crdb.dev/check/cockroachdb/cockroach?pullRequest=105764) it.</sub>
